bundle/terranova/deploy_mutator.go

```diff
@@ -0,0 +1,255 @@
+
+func TerranovaDeploy() bundle.Mutator {
+	return &terranovaDeployMutator{}
+}
+
+func (m *terranovaDeployMutator) Name() string {
+	return "TerranovaDeploy"
+}
+
+func (m *terranovaDeployMutator) Apply(ctx context.Context, b *bundle.Bundle) diag.Diagnostics {
+	client := b.WorkspaceClient()
+
+	g := dagrun.NewGraph[tnstate.ResourceNode]()
+
+	_, err := dyn.MapByPattern(
+		b.Config.Value(),
+		dyn.NewPattern(dyn.Key("resources"), dyn.AnyKey(), dyn.AnyKey()),
+		func(p dyn.Path, v dyn.Value) (dyn.Value, error) {
+			section := p[1].Key()
+			name := p[2].Key()
+			g.AddNode(tnstate.ResourceNode{
+				Section: section,
+				Name:    name,
+			})
+
+			// TODO: Scan v for references and use g.AddDirectedEdge to add dependency
+			return v, nil
+		},
+	)
+
+	if g.Size() == 0 {
+		// Avoid creating state file if nothing to deploy
+		return nil
+	}
+
+	var diags diag.SafeDiagnostics
+
+	err = g.Run(defaultParallelism, func(node tnstate.ResourceNode) {
+		// TODO: if a given node fails, all downstream nodes should not be run. We should report those nodes.
+		// TODO: ensure that config for this node is fully resolved at this point.
+
+		config, ok := b.GetResourceConfig(node.Section, node.Name)
+		if !ok {
+			diags.AppendErrorf("internal error: cannot get config for %s", node)
+			return
+		}
+
+		d := Deployer{
+			client:       client,
+			db:           &b.ResourceDatabase,
+			section:      node.Section,
+			resourceName: node.Name,
+		}
+
+		err = d.Deploy(ctx, config)
+		if err != nil {
+			diags.AppendError(err)
+			return
+		}
+	})
+	if err != nil {
+		diags.AppendError(err)
+	}
+
+	err = b.ResourceDatabase.Finalize()
+	if err != nil {
+		diags.AppendError(err)
+	}
+
+	return diags.Diags
+}
+
+type Deployer struct {
+	client       *databricks.WorkspaceClient
+	db           *tnstate.TerranovaState
+	section      string
+	resourceName string
+}
+
+func (d *Deployer) Deploy(ctx context.Context, inputConfig any) error {
+	err := d.deploy(ctx, inputConfig)
+	if err != nil {
+		return fmt.Errorf("deploying %s.%s: %w", d.section, d.resourceName, err)
+	}
+	return nil
+}
+
+func (d *Deployer) deploy(ctx context.Context, inputConfig any) error {
+	entry, hasEntry := d.db.GetResourceEntry(d.section, d.resourceName)
+
+	resource, err := tnresources.New(d.client, d.section, d.resourceName, inputConfig)
+	if err != nil {
+		return err
+	}
+
+	config := resource.Config()
+
+	if !hasEntry {
+		return d.Create(ctx, resource, config)
+	}
+
+	oldID := entry.ID
+	if oldID == "" {
+		return errors.New("invalid state: empty id")
+	}
+
+	savedState, err := typeConvert(resource.GetType(), entry.State)
+	if err != nil {
+		return fmt.Errorf("interpreting state: %w", err)
+	}
+
+	localDiff, err := structdiff.GetStructDiff(savedState, config)
+	if err != nil {
+		return fmt.Errorf("comparing state and config: %w", err)
+	}
+
+	localDiffType := tnresources.ChangeTypeNone
+	if len(localDiff) > 0 {
+		localDiffType = resource.ClassifyChanges(localDiff)
```

Comment:
What happens if the resource is removed remotely? We calculate the change here based on only local state and diff, right?

Reply:
Right. Please review again. Updated version creates a plan first with deletions tracked.

Reply:
Update: the latest version does create a plan but it still not tracking deletions in the config. This will be addressed in a follow up.